rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,239 @@
+  be left for future improvement.
+
+- *Principal*
+  The principal is the entity that originates the event. This field is mandatory
+  and can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Src*
+  The source entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Target*
+  The target entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+### Implementation
+
+Both event types require different batching strategies.
+
+#### UDM
+
+UDM can use a simple `BatchedHttpSink` with a `JsonArrayBuffer` to perform the
+batching.
+
+#### Unstructured log events
+
+Since log events are batched by `log_type` we will need to partition the
+batches on the value of the `log_type` field using a `PartitionHttpSink`.
+
+## Rationale
+
+There is significant user demand for sending data into Google Chronicle.
+
+## Drawbacks
+
+Adding a new sink does place an additional maintenance burden on the team. In
+particular since Chronicle is a third party service, developing integration
+tests to ensure the sink continues to work correctly is burdensome.
+
+## Prior Art
+
+## Alternatives
+
+An option that was considered was to use the HTTP sink and and enhance it to
+allow GCP authentication. This would nearly work for UDM events, but
+unfortunately since the payload is a JSON object with a `customer_id` field
+and an array of `events`, the structure is slightly different to what is
+possible with the HTTP sink.
+
+## Outstanding Questions
+
+Should we crate one sink with an `event_type` option, or should these be two
+separate sinks?
+
+The `log_types` field has to contain a value that has been returned from the
+`log_types` endpoint. It could be possible for the Sink to call this on boot
+time which would provide a list of possible values that could be validated
+before sending to chronicle. I feel it isn't worth doing this as the event
+will be validated by Chronicle anyway so we may as well rely on the error
+returned when sending an invalid `log_type`. Others may disagree. Is it worth
+performing this validation ourselves?
+
+We would need to take into account that Vector would then need rebooting
+should the `log_types` list be updated.
+
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues
+after the RFC is approved:
+
+- [ ] Create the sink to work for unstructured log events.
+- [ ] Enhance the sink to work with UDM (without schema support).
+- [ ] Add schema support for UDM events.
+
+## Future Improvements
+
+- Consider sending the event data as Proto3 rather than JSON. We would likely
```

Comment:
I did a quick search for the .proto file(s), I did not find them, do you know if they are freely available for reuse ?

Reply:
I will find out.